In [1]:
import pandas as pd
import numpy as np

# Index

1. Exploratory Data Analysis

## 1. Exploratory Data Analysis

How many categorical and numerical columns we have?

In [2]:
df=pd.read_csv("Employee_Attrition.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       1323 non-null   float64
 1   Attrition                 1323 non-null   object 
 2   BusinessTravel            1323 non-null   object 
 3   DailyRate                 1323 non-null   float64
 4   Department                1323 non-null   object 
 5   DistanceFromHome          1323 non-null   float64
 6   Education                 1323 non-null   float64
 7   EducationField            1323 non-null   object 
 8   EmployeeCount             1323 non-null   float64
 9   EmployeeNumber            1323 non-null   float64
 10  EnvironmentSatisfaction   1323 non-null   float64
 11  Gender                    1323 non-null   object 
 12  HourlyRate                1323 non-null   float64
 13  JobInvolvement            1323 non-null   float64
 14  JobLevel

As we can see:
    We have 1470 entries in total and 147 with missing values (1470-1323).
    We have 9 categorical columns and 26 numerical ones. Ours datatypes are float and objects.

Columns to perform the study based on "JobSatisfaction" and "PerformanceRating"

Entries with nun values

In [3]:
df.isnull().sum()

Age                         147
Attrition                   147
BusinessTravel              147
DailyRate                   147
Department                  147
DistanceFromHome            147
Education                   147
EducationField              147
EmployeeCount               147
EmployeeNumber              147
EnvironmentSatisfaction     147
Gender                      147
HourlyRate                  147
JobInvolvement              147
JobLevel                    147
JobRole                     147
JobSatisfaction             147
MaritalStatus               147
MonthlyIncome               147
MonthlyRate                 147
NumCompaniesWorked          147
Over18                      147
OverTime                    147
PercentSalaryHike           147
PerformanceRating           147
RelationshipSatisfaction    147
StandardHours               147
StockOptionLevel            147
TotalWorkingYears           147
TrainingTimesLastYear       147
WorkLifeBalance             147
YearsAtC

In [4]:
df.isnull().sum().sum()

5145

As we can see every column has at least 147 missing values

Analysis of missing values

The first thing I would like to get to know if whether or not our nan values are float and np.nan or strings "nan".

In [5]:
# Assuming df is your DataFrame
for column in df.columns:
    # Check if there are NaN values in the column
    if df[column].isnull().any():
        # Extract the first non-null value for type comparison
        non_null_value = df[column].dropna().iloc[0]

        # Check if the non-null value is of type float
        if isinstance(non_null_value, float):
            # If it's a float, then NaNs are likely numpy.nan (float)
            print(f"Column '{column}': NaN values are of type float (numpy.nan)")
        else:
            # If it's not a float, then NaNs are likely of type string
            print(f"Column '{column}': NaN values are likely of type string ('nan')")

Column 'Age': NaN values are of type float (numpy.nan)
Column 'Attrition': NaN values are likely of type string ('nan')
Column 'BusinessTravel': NaN values are likely of type string ('nan')
Column 'DailyRate': NaN values are of type float (numpy.nan)
Column 'Department': NaN values are likely of type string ('nan')
Column 'DistanceFromHome': NaN values are of type float (numpy.nan)
Column 'Education': NaN values are of type float (numpy.nan)
Column 'EducationField': NaN values are likely of type string ('nan')
Column 'EmployeeCount': NaN values are of type float (numpy.nan)
Column 'EmployeeNumber': NaN values are of type float (numpy.nan)
Column 'EnvironmentSatisfaction': NaN values are of type float (numpy.nan)
Column 'Gender': NaN values are likely of type string ('nan')
Column 'HourlyRate': NaN values are of type float (numpy.nan)
Column 'JobInvolvement': NaN values are of type float (numpy.nan)
Column 'JobLevel': NaN values are of type float (numpy.nan)
Column 'JobRole': NaN values

In [6]:
df= df.replace(['nan'],[np.NaN])

In [7]:
df.describe(include='all')

Age Attrition BusinessTravel    DailyRate  \
count   1323.000000      1323           1323  1323.000000   
unique          NaN         2              3          NaN   
top             NaN        No  Travel_Rarely          NaN   
freq            NaN      1065            890          NaN   
mean      36.635268       NaN            NaN   802.030050   
std        9.882161       NaN            NaN   414.031025   
min       14.544608       NaN            NaN    86.828045   
25%       29.624982       NaN            NaN   456.479553   
50%       34.951950       NaN            NaN   782.958705   
75%       42.748741       NaN            NaN  1130.577949   
max       69.402515       NaN            NaN  1784.394456   

                    Department  DistanceFromHome    Education EducationField  \
count                     1323       1323.000000  1323.000000           1323   
unique                       3               NaN          NaN              6   
top     Research & Development               NaN          NaN  Life Sciences   
freq                       824               NaN          NaN            504   
mean                       NaN          9.091097     2.923447            NaN   
std                        NaN          8.177870     1.072203            NaN   
min                        NaN          0.750112     0.770362            NaN   
25%                        NaN          2.147069     2.122706            NaN   
50%                        NaN          6.920808     3.015658            NaN   
75%                        NaN         13.584900     3.700560            NaN   
max                        NaN         33.684424     6.285161            NaN   

        EmployeeCount  EmployeeNumber  ...  RelationshipSatisfaction  \
count     1323.000000     1323.000000  ...               1323.000000   
unique            NaN             NaN  ...                       NaN   
top               NaN             NaN  ...                       NaN   
freq              NaN             NaN  ...                       NaN   
mean         0.999325     1026.679166  ...                  2.720431   
std          0.099656      610.480802  ...                  1.115959   
min          0.628985        1.054148  ...                  0.720696   
25%          0.931146      501.854842  ...                  1.909010   
50%          0.998949     1022.025017  ...                  2.891235   
75%          1.067371     1533.811096  ...                  3.639961   
max          1.290613     2530.919375  ...                  5.092280   

       StandardHours  StockOptionLevel  TotalWorkingYears  \
count    1323.000000       1323.000000        1323.000000   
unique           NaN               NaN                NaN   
top              NaN               NaN                NaN   
freq             NaN               NaN                NaN   
mean       80.241313          0.814757          11.426152   
std         8.022535          0.863076           8.075870   
min        53.793059          0.000000           0.000000   
25%        74.882946          0.000000           5.953221   
50%        80.351400          0.913651           9.306742   
75%        85.664169          1.082523          15.324548   
max       107.378705          3.474319          47.641892   

        TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
count             1323.000000     1323.000000     1323.000000   
unique                    NaN             NaN             NaN   
top                       NaN             NaN             NaN   
freq                      NaN             NaN             NaN   
mean                 2.811756        2.762825        6.933624   
std                  1.329044        0.761802        6.046798   
min                  0.000000        0.799138        0.000000   
25%                  1.967328        2.193390        2.673269   
50%                  2.690934        2.899807        5.422160   
75%                  3.366949        3.206117        9.238005   
max      

Statisitical description of categorical columns?

In [8]:
df.describe(include="O")

Attrition BusinessTravel              Department EducationField Gender  \
count       1323           1323                    1323           1323   1323   
unique         2              3                       3              6      2   
top           No  Travel_Rarely  Research & Development  Life Sciences   Male   
freq        1065            890                     824            504    783   

                JobRole MaritalStatus Over18 OverTime  
count              1323          1323   1323     1323  
unique                9             3      1        2  
top     Sales Executive       Married      Y       No  
freq                281           596   1323      929

In [9]:
df['BusinessTravel'].value_counts()

Travel_Rarely        890
Travel_Frequently    279
Non-Travel           154
Name: BusinessTravel, dtype: int64

In [10]:
df['OverTime'].value_counts()

No     929
Yes    394
Name: OverTime, dtype: int64

In [11]:
df['Attrition'].value_counts()

No     1065
Yes     258
Name: Attrition, dtype: int64

In [12]:
df['Department'].value_counts()

Research & Development    824
Sales                     393
Human Resources           106
Name: Department, dtype: int64

In [13]:
df['EducationField'].value_counts()

Life Sciences       504
Medical             397
Marketing           149
Technical Degree    141
Other                83
Human Resources      49
Name: EducationField, dtype: int64

In [14]:
df['Gender'].value_counts()

Male      783
Female    540
Name: Gender, dtype: int64

In [15]:
df['JobRole'].value_counts()

Sales Executive              281
Research Scientist           257
Laboratory Technician        229
Manufacturing Director       122
Healthcare Representative    118
Manager                      101
Research Director             80
Sales Representative          75
Human Resources               60
Name: JobRole, dtype: int64

In [16]:
df['Over18'].value_counts()

Y    1323
Name: Over18, dtype: int64

In [17]:
df['MaritalStatus'].value_counts()

Married     596
Single      429
Divorced    298
Name: MaritalStatus, dtype: int64

Statisitical description of numerical columns?

In [18]:
df.describe()

Age    DailyRate  DistanceFromHome    Education  EmployeeCount  \
count  1323.000000  1323.000000       1323.000000  1323.000000    1323.000000   
mean     36.635268   802.030050          9.091097     2.923447       0.999325   
std       9.882161   414.031025          8.177870     1.072203       0.099656   
min      14.544608    86.828045          0.750112     0.770362       0.628985   
25%      29.624982   456.479553          2.147069     2.122706       0.931146   
50%      34.951950   782.958705          6.920808     3.015658       0.998949   
75%      42.748741  1130.577949         13.584900     3.700560       1.067371   
max      69.402515  1784.394456         33.684424     6.285161       1.290613   

       EmployeeNumber  EnvironmentSatisfaction   HourlyRate  JobInvolvement  \
count     1323.000000              1323.000000  1323.000000     1323.000000   
mean      1026.679166                 2.726443    65.647722        2.743028   
std        610.480802                 1.129679    21.469709        0.765568   
min          1.054148                 0.759655    23.855272        0.722008   
25%        501.854842                 1.894818    47.701897        2.177638   
50%       1022.025017                 2.887313    64.934490        2.859143   
75%       1533.811096                 3.676600    83.038355        3.179128   
max       2530.919375                 5.254672   123.200439        5.241968   

          JobLevel  ...  RelationshipSatisfaction  StandardHours  \
count  1323.000000  ...               1323.000000    1323.000000   
mean      2.058270  ...                  2.720431      80.241313   
std       1.127018  ...                  1.115959       8.022535   
min       0.715172  ...                  0.720696      53.793059   
25%       1.045228  ...                  1.909010      74.882946   
50%       1.936150  ...                  2.891235      80.351400   
75%       2.609707  ...                  3.639961      85.664169   
max       5.942269  ...                  5.092280     107.378705   

       StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
count       1323.000000        1323.000000            1323.000000   
mean           0.814757          11.426152               2.811756   
std            0.863076           8.075870               1.329044   
min            0.000000           0.000000               0.000000   
25%            0.000000           5.953221               1.967328   
50%            0.913651           9.306742               2.690934   
75%            1.082523          15.324548               3.366949   
max            3.474319          47.641892               7.503201   

       WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
count      1323.000000     1323.000000         1323.000000   
mean          2.762825        6.933624            4.233363   
std           0.761802        6.046798            3.658111   
min           0.799138        0.000000            0.000000   
25%           2.193390        2.673269            1.824392   
50%           2.899807        5.422160            2.968874   
75%           3.206117        9.238005            7.052591   
max           4.856086       36.848408           19.091844   

       YearsSinceLastPromotion  YearsWithCurrManager  
count              1323.000000           1323.000000  
mean                  2.160710              4.117695  
std                   3.221481              3.594053  
min                   0.000000              0.000000  
25%                   0.000000              1.687177  
50%                   0.967140              2.939240  
75%                   2.586631              6.987964  
max                  17.273662             19.976003  

[8 rows x 26 columns]

References

Stack Overflow. (n.d.). Pandas not recognizing NaN as null. [online] Available at: https://stackoverflow.com/questions/30604893/pandas-not-recognizing-nan-as-null [Accessed 28 Dec. 2023].

pandas.pydata.org. (n.d.). pandas.isnull — pandas 1.4.2 documentation. [online] Available at: https://pandas.pydata.org/docs/reference/api/pandas.isnull.html.

note.nkmk.me. (2023). pandas: Find rows/columns with NaN (missing values) | note.nkmk.me. [online] Available at: https://note.nkmk.me/en/python-pandas-nan-extract/ [Accessed 28 Dec. 2023].

saturncloud.io. (2023). How to Check if a Single Cell Value is NaN in Pandas | Saturn Cloud Blog. [online] Available at: https://saturncloud.io/blog/how-to-check-if-a-single-cell-value-is-nan-in-pandas/#:~:text=One%20of%20the%20most%20common [Accessed 28 Dec. 2023].

Vanawat, N. (2021). How To Perform Exploratory Data Analysis -A Guide for Beginners. [online] Analytics Vidhya. Available at: https://www.analyticsvidhya.com/blog/2021/08/how-to-perform-exploratory-data-analysis-a-guide-for-beginners/.
    